### PR TITLE
integration-test: remove bpftool dependency

### PR DESCRIPTION
**Pull request by tamird**
- sys: `sys_bpf` takes mut ref
- sys: extract common SyscallError
- programs: make `loaded_programs` opaque
- sys: push error construction up
- programs: plug file descriptor leak
- Add links iterator

- - -
This change is [Reviewable](https://reviewable.io/reviews/aya-rs/aya/712)

### CODE REVIEW

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r1, 40 of 40 files at r2, 2 of 2 files at r3, 10 of 10 files at r4, 1 of 1 files at r5, 5 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/712#-:-NamVsK44A-uE_m9cCp1:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @alessandrod, @astoycos, @dave-tucker, and @tamird)

___
*[`test/integration-test/src/tests/load.rs` line 88 at r6](https://reviewable.io/reviews/aya-rs/aya/712#-NamVgcVBS3itZGKlIVF:-NamVgcVBS3itZGKlIVG:b-3j1btj) ([raw file](https://github.com/aya-rs/aya/blob/f08a79149180a3c7ca7e896a9a865dc161234db9/test/integration-test/src/tests/load.rs#L88)):*
> ```XML
>         thread::sleep(RETRY_DURATION);
>     }
>     panic!("exhausted {MAX_RETRIES} attempts, wanted loaded={loaded}");
> ```

Is using `track_caller` here enough to get something useful when this fails? I recall in the past needing to explicitly use the caller info, maybe add that?

**Review by tamird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/712)* status: 30 of 44 files reviewed, all discussions resolved (waiting on @ajwerner, @alessandrod, @astoycos, and @dave-tucker)

___
*[`test/integration-test/src/tests/load.rs` line 88 at r6](https://reviewable.io/reviews/aya-rs/aya/712#-NamVgcVBS3itZGKlIVF:-NamYZfk3BAeb3svloEv:b-x8kjq) ([raw file](https://github.com/aya-rs/aya/blob/f08a79149180a3c7ca7e896a9a865dc161234db9/test/integration-test/src/tests/load.rs#L88)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Is using `track_caller` here enough to get something useful when this fails? I recall in the past needing to explicitly use the caller info, maybe add that?
</blockquote></details>

It's not necessary, panic already internally does that. What you're recalling was a manual construction we used to plumb line numbers around (without panicking).

**Review by dave-tucker — COMMENTED** [review]
> Reviewed 2 of 2 files at r1, 30 of 40 files at r2, 10 of 14 files at r7, 1 of 1 files at r8, 1 of 3 files at r9, 8 of 10 files at r10, 4 of 5 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/712)* status: 43 of 44 files reviewed, 5 unresolved discussions (waiting on @ajwerner, @alessandrod, @astoycos, and @tamird)

___
*[`aya/src/pin.rs` line 3 at r8](https://reviewable.io/reviews/aya-rs/aya/712#-Nan35L11QSGiUBffCwA:-Nan35L11QSGiUBffCwB:b3tqnwi) ([raw file](https://github.com/aya-rs/aya/blob/c1718af2d8cf602e3aa2140e9a92d2ba5dd333c6/aya/src/pin.rs#L3)):*
> ```XML
> //! Pinning BPF objects to the BPF filesystem.
> 
> use crate::sys::SyscallError;
> ```

Did we ever tell clippy to do the import ordering?
We've been trying to use std, deps, crate but I'm not sure that's enforced (yet)
___
*[`aya/src/programs/mod.rs` line 1003 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-Nan6pCT0ZO-Plt-Gaen:-Nan6pCT0ZO-Plt-Gaeo:b-orgx8v) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/aya/src/programs/mod.rs#L1003)):*
> ```XML
> }
> 
> // TODO(https://github.com/aya-rs/aya/issues/645): add documentation.
> ```

It's not just documentation, I don't think we want to leak bpf_link_info into the public API.
At a minimum it would be a tuple struct with accessors i.e `pub struct LinkInfo(bpf_link_info)`.
If it's not too much bother, you could add them in this patch series.
___
*[`aya/src/sys/bpf.rs` line 512 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-Nan9bFL-kLLP51abtfK:-Nan9bFL-kLLP51abtfL:bwh8e8h) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/aya/src/sys/bpf.rs#L512)):*
> ```XML
> 
> pub(crate) fn bpf_link_get_fd_by_id(link_id: u32) -> Result<OwnedFd, SyscallError> {
>     let mut attr = unsafe { mem::zeroed::<bpf_attr>() };
> ```

Should we be using MaybeUninit here?
___
*[`test/integration-test/src/tests/load.rs` line 72 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-NanBGYNBGDdWEKK683V:-NanBGYNBGDdWEKK683W:bhvggt3) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/test/integration-test/src/tests/load.rs#L72)):*
> ```XML
>     }
> 
>     panic!("exhausted {MAX_RETRIES} attempts, wanted loaded_and_linked={loaded_and_linked}");
> ```

It would be helpful here to know which of the 2 conditions weren't met
___
*[`test/integration-test/src/tests/load.rs` line 75 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-NanCAmp2nb0ce3PG7Au:-NanCAmp2nb0ce3PG7Av:bmnrwuf) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/test/integration-test/src/tests/load.rs#L75)):*
> ```XML
> }
> 
> #[track_caller]
> ```

How does this change the behaviour of how panics in tests are displayed?

**Review by tamird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/712)* status: 43 of 44 files reviewed, 2 unresolved discussions (waiting on @ajwerner, @alessandrod, @astoycos, and @dave-tucker)

___
*[`aya/src/pin.rs` line 3 at r8](https://reviewable.io/reviews/aya-rs/aya/712#-Nan35L11QSGiUBffCwA:-NanD27RDuFtlJKp5F4M:b5kijrh) ([raw file](https://github.com/aya-rs/aya/blob/c1718af2d8cf602e3aa2140e9a92d2ba5dd333c6/aya/src/pin.rs#L3)):*
<details><summary><i>Previously, dave-tucker (Dave Tucker) wrote…</i></summary><blockquote>

Did we ever tell clippy to do the import ordering?
We've been trying to use std, deps, crate but I'm not sure that's enforced (yet)
</blockquote></details>

Given that this PR is green, it's not enforced. I'd rather not try to divine that here, particularly if we can get the robots to do it.
___
*[`aya/src/programs/mod.rs` line 1003 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-Nan6pCT0ZO-Plt-Gaen:-NanDIppCbFogAI_pbVZ:b-ccbud7) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/aya/src/programs/mod.rs#L1003)):*
<details><summary><i>Previously, dave-tucker (Dave Tucker) wrote…</i></summary><blockquote>

It's not just documentation, I don't think we want to leak bpf_link_info into the public API.
At a minimum it would be a tuple struct with accessors i.e `pub struct LinkInfo(bpf_link_info)`.
If it's not too much bother, you could add them in this patch series.
</blockquote></details>

Is that important to note here? I'd rather not golf on the API here because of how long that tends to take. @astoycos volunteered, as I recall.
___
*[`aya/src/sys/bpf.rs` line 512 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-Nan9bFL-kLLP51abtfK:-NanDXDpEDyqRWr_gl9B:bhl81pf) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/aya/src/sys/bpf.rs#L512)):*
<details><summary><i>Previously, dave-tucker (Dave Tucker) wrote…</i></summary><blockquote>

Should we be using MaybeUninit here?
</blockquote></details>

No. We are initializing this.
___
*[`test/integration-test/src/tests/load.rs` line 72 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-NanBGYNBGDdWEKK683V:-NanKOCm30xPpPslhnjM:b-h6jauc) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/test/integration-test/src/tests/load.rs#L72)):*
<details><summary><i>Previously, dave-tucker (Dave Tucker) wrote…</i></summary><blockquote>

It would be helpful here to know which of the 2 conditions weren't met
</blockquote></details>

Done. This now only polls the program until it's loaded and then just looks for the link.
___
*[`test/integration-test/src/tests/load.rs` line 75 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-NanCAmp2nb0ce3PG7Au:-NanDpRn6wcF2NnoWXcu:bykxfck) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/test/integration-test/src/tests/load.rs#L75)):*
<details><summary><i>Previously, dave-tucker (Dave Tucker) wrote…</i></summary><blockquote>

How does this change the behaviour of how panics in tests are displayed?
</blockquote></details>

They'll use the caller's line number.

**Review by dave-tucker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/712)* status: 0 of 44 files reviewed, 1 unresolved discussion (waiting on @ajwerner, @alessandrod, and @astoycos)

___
*[`aya/src/programs/mod.rs` line 1003 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-Nan6pCT0ZO-Plt-Gaen:-NanPFu9ENo4cDQKvYg-:b-7maduc) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/aya/src/programs/mod.rs#L1003)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Is that important to note here? I'd rather not golf on the API here because of how long that tends to take. @astoycos volunteered, as I recall.
</blockquote></details>

The thing is, that TODO could be interpreted as  "I can remove `#[doc(hidden)]` if I add documentation"
I would rather it said something like:
```
// TODO: This API was added for testing.
// Stabilize this API before removing doc(hidden).
// Design discussion: https://github.com/aya-rs/aya/issues/645
```
___
*[`aya/src/sys/bpf.rs` line 512 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-Nan9bFL-kLLP51abtfK:-NanSrl37AilBZCCq-oE:bx9awhu) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/aya/src/sys/bpf.rs#L512)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

No. We are initializing this.
</blockquote></details>

Sorry I should have been more explicit. I meant`MaybeUninit::zerored().assume_init()`

**Review by astoycos — COMMENTED** [review]
Still have more to review here but I have to step away for now 👍

File: aya/src/sys/bpf.rs [hunk 1]
```diff
@@ -460,57 +460,71 @@ pub(crate) fn bpf_prog_query(
...
+        assert_eq!(code, -1);
+        SyscallError {
+            call: "bpf_prog_get_fd_by_id",
+            io_error,
+        }
+    })
 }
 
-pub(crate) fn bpf_prog_get_info_by_fd(prog_fd: RawFd) -> Result<bpf_prog_info, io::Error> {
+fn bpf_obj_get_info_by_fd<T>(fd: BorrowedFd<'_>) -> Result<T, SyscallError> {
```
[the remark this answers is not in the document]
Comment (tamird) [reply aya/src/sys/bpf.rs:477]: We can massage it, or simply not use the generic. Do we need to solve that here?

File: aya/src/sys/bpf.rs [hunk 2]
```diff
@@ -990,21 +1004,63 @@ pub(crate) fn is_btf_type_tag_supported() -> bool {
...
     let mut attr = unsafe { mem::zeroed::<bpf_attr>() };
     let u = unsafe { &mut attr.__bindgen_anon_6 };
     u.__bindgen_anon_1.start_id = id;
-    match sys_bpf(bpf_cmd::BPF_PROG_GET_NEXT_ID, &attr) {
-        Ok(_) => Ok(Some(unsafe { attr.__bindgen_anon_6.next_id })),
-        Err((_, io_error)) if io_error.raw_os_error() == Some(ENOENT) => Ok(None),
-        Err(e) => Err(e),
+    match sys_bpf(cmd, &mut attr) {
+        Ok(code) => {
+            assert_eq!(code, 0);
```
[the remark this answers is not in the document]
Comment (tamird) [reply aya/src/sys/bpf.rs:936]: That matches any value greater than zero, this asserts precisely zero.

**Review by tamird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/712)* status: 0 of 44 files reviewed, 3 unresolved discussions (waiting on @ajwerner, @alessandrod, @astoycos, and @dave-tucker)

___
*[`aya/src/programs/mod.rs` line 1003 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-Nan6pCT0ZO-Plt-Gaen:-NanearYBlNMuOlaIQsH:b-fjtf4o) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/aya/src/programs/mod.rs#L1003)):*
<details><summary><i>Previously, dave-tucker (Dave Tucker) wrote…</i></summary><blockquote>

The thing is, that TODO could be interpreted as  "I can remove `#[doc(hidden)]` if I add documentation"
I would rather it said something like:
```
// TODO: This API was added for testing.
// Stabilize this API before removing doc(hidden).
// Design discussion: https://github.com/aya-rs/aya/issues/645
```
</blockquote></details>

Works for me. Thanks for the concrete suggestion.
___
*[`aya/src/sys/bpf.rs` line 512 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-Nan9bFL-kLLP51abtfK:-NanevJ-6l-awl6K3Ly9:b-9knt6e) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/aya/src/sys/bpf.rs#L512)):*
<details><summary><i>Previously, dave-tucker (Dave Tucker) wrote…</i></summary><blockquote>

Sorry I should have been more explicit. I meant`MaybeUninit::zerored().assume_init()`
</blockquote></details>

How is that different from this? How is it better?

**Review by alessandrod — APPROVED** [review]

**Review by dave-tucker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/712)* status: 0 of 44 files reviewed, 2 unresolved discussions (waiting on @ajwerner and @astoycos)

___
*[`aya/src/sys/bpf.rs` line 512 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-Nan9bFL-kLLP51abtfK:-NaqBKbt3eUPkBEJ_bYQ:b-x22ycu) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/aya/src/sys/bpf.rs#L512)):*
I'm not sure - this is a genuine question - and something that's been on my mind for a while... it's not strictly necessary for this PR, but you've triggered me to ask :)

From the docs for mem::zeroed: https://doc.rust-lang.org/std/mem/fn.zeroed.html

> This has the same effect as [`MaybeUninit::zeroed().assume_init()`](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.zeroed "associated function std::mem::MaybeUninit::zeroed"). It is useful for FFI sometimes, but should generally be avoided.

Same code with MaybeUninit

```rust 
let attr = MaybeUninit::<bpf_attr>::zeroed();
// Safety: bpf_attr has been zeroed so it's safe to assume init
let attr = unsafe { attr.assume_init() };
```

Honestly not sure what's better.

**Review by tamird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/712)* status: 0 of 44 files reviewed, 2 unresolved discussions (waiting on @ajwerner, @astoycos, and @dave-tucker)

___
*[`aya/src/sys/bpf.rs` line 512 at r12](https://reviewable.io/reviews/aya-rs/aya/712#-Nan9bFL-kLLP51abtfK:-NaqE1Zd8QHvlyJ-zLQ5:b-2n67jf) ([raw file](https://github.com/aya-rs/aya/blob/8a48545694fd563ac60384212279f25367681ccf/aya/src/sys/bpf.rs#L512)):*
<details><summary><i>Previously, dave-tucker (Dave Tucker) wrote…</i></summary><blockquote>

I'm not sure - this is a genuine question - and something that's been on my mind for a while... it's not strictly necessary for this PR, but you've triggered me to ask :)

From the docs for mem::zeroed: https://doc.rust-lang.org/std/mem/fn.zeroed.html

> This has the same effect as [`MaybeUninit::zeroed().assume_init()`](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.zeroed "associated function std::mem::MaybeUninit::zeroed"). It is useful for FFI sometimes, but should generally be avoided.

Same code with MaybeUninit

```rust 
let attr = MaybeUninit::<bpf_attr>::zeroed();
// Safety: bpf_attr has been zeroed so it's safe to assume init
let attr = unsafe { attr.assume_init() };
```

Honestly not sure what's better.
</blockquote></details>

Yeah, that phrasing reads like someone is hoping to remove mem::zeroed someday in the distant future just so there are fewer ways to do the same thing.

I'd be supportive of a PR that replaced all uses of `mem::zeroed` just for consistency. In the interest of not acquiring the `git blame` and eating more CI cycles I'm going to resolve this thread here.

File: aya/src/sys/bpf.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (astoycos) [reply aya/src/sys/bpf.rs:477]: No I'll just have to solve it in https://github.com/aya-rs/aya/pull/637

**Pull request merged**